# КОММЕНТАРИЙ РЕВЬЮВЕРА 2:

Приветствую тебя еще раз!) Спасибо за то, что доработал проект!

Мои комментарии будут также обозначены разными цветами, но с пометкой 2:
    
    «КОММЕНТАРИЙ РЕВЬЮВЕРА 2: ... »

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 

Сергей, ты хорошо доработал проект, думаю теперь я могу его принять, желаю тебе успехов в дальнейшем обучении!)

</div>

# КОММЕНТАРИЙ ОТ РЕВЬЮВЕРА
Привет, Сергей!)

Меня зовут Обидин Андрей, я проверял твой проект. Благодарю тебя за то, что вовремя всё сдал и хорошо поработал!) Мне понраивлся твой проект, однако есть несколько замечаний и рекомендаций, пожалуйста поработай с ними. После я смогу принять твой проект!) В конце есть чек-лист

Далее по ходу работы ты сможешь встретить мои комментарии, они будут отмечены разными цветами:

<div class="alert alert-block alert-success">
Похвала и возможности развития
</div>
<div class="alert alert-block alert-warning">
Некритичные замечания, но стоит обратить внимание
</div>
<div class="alert alert-block alert-danger">
Грубая ошибка, необходимо исправить
</div>


Заранее прошу не удалять мои комментарии, это важно. Если у тебя возникнут вопросы, обозначь их, я с радостью тебе помогу!)

Как вариант, можешь обозначить их так:
<div class="alert alert-block alert-info">
<b>КОММЕНТАРИЙ СТУДЕНТА:</b>
</div>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

В начале проекта стоит давать чуть более развернутое описание, чтобы люди, которые будут его смотреть, могли полноценно со всем ознакомиться. Помимо краткой цели проекта рекомендую дать описание признаков данных и небольшое содержание. Можешь даже сделать <a href="https://sebastianraschka.com/Articles/2014_ipython_internal_links.html">кликабельные разделы</a>, чтобы ориентироваться в проекте было еще проще.
</div>

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [26]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv') #импортируем и читаем файл базы данных
print(data.info()) #выводим информацию о базе данных. Видим общее количество строк 21525,пропуски есть в графах: 'total_income', 
#'days_employed'. 
print('Посмотрим на первые 10 строк нашей таблицы:')
print(data.head(10))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None
Посмотрим на первые 10 строк нашей таблицы:
   children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124

### Вывод

В нашем распоряжении таблица из 12 столбцов и 21525 строк. Пропуски есть в столбцах days_employed и total_income, эти же столбцы имеют тип данных float64. 

<div class="alert alert-block alert-danger">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Выведи начало датафрейма
    
</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 

Молодец, всегда стоит осматривать датафрейм (можно использовать head \ tail \ sample)

</div>

### Шаг 2. Предобработка данных

In [27]:

print('Ищем пропуски в столбцах')
print(data.isnull().sum()) #убеждаемся в предыдущей гипотезе, вызвав метод isnull и подсчитав количество пропусков. Пропуски 
# данные графы содержат в себе количественные переменные, значит заполнять их будем характерными значениями. Природа пропусков
# в графе с доходом неясна, так как это обязательная входящая информация, которую сотрудники банка должны проверять, прежде, чем
# выдавать кредит. Вместе с этим проверим, есть ли отрицательные значения в графе с доходом.


print()
print()
print('Убеждаемся в отсутствии минусовых значений зарплат')
print(data['total_income'][data['total_income']<0].count()) #отлично отрицательных значений нет. Значит все участники базы с указанным доходом, что-то
#да зарабатывают.Люди в разном возрасте зарабатывают разные деньги, поэтому посмотрим, какой возраст у наших будщих должников.


print() #делаем пару пробелов, чтобы выводы не сливались в одно целое.
print()
#Для этого напишем функцию, которая будет принимать возрастной интервал и выводить, сколько клиентов данного интервала имело
#историю с кредитами.
def how_old_count(age1,age2):
    return(data['dob_years'][(data['dob_years']>=age1) & (data['dob_years']<=age2)].count())
def miss_count(age1,age2):
    return(data['total_income'][(data['dob_years']>=age1) & (data['dob_years']<=age2)].isnull().sum())
#Выводим данные по количеству заемщиков, чтобы данные сохранились и можно к ним было обращаться, сделаем DataFrame
#dob_years_count + посчитаем сразу количество пропусков в этих же интервалах.


how_old_data = [
    ['1 до 20 лет:', how_old_count(1,20), miss_count(1,20)],
    ['21 до 30 лет:',how_old_count(21,30), miss_count(21,30)],
    ['31 до 40 лет:',how_old_count(31,40), miss_count(31,40)],
    ['от 41 до 50 лет:',how_old_count(41,50), miss_count(41,50)],
    ['от 51 до 60 лет:',how_old_count(51,60), miss_count(51,60)],
    ['от 61 до 70 лет:',how_old_count(61,70), miss_count(61,70)],
    ['от 71 до 100 лет:',how_old_count(71,100), miss_count(71,100)],
    ['от 100 до 200 лет:',how_old_count(100,200), miss_count(100,200)]
]#DataFrame, который на всякий случай будет хранить данные по возратам.

columns = ['Возраст заемщиков', 'Их количество','Количество пропусков']
dob_years_count = pd.DataFrame(data = how_old_data, columns = columns)
print('Считаем, сколько заемщиков берут средства в разных интервалах')                               
print(dob_years_count)
print()    

#Как можно увидеть из таблицы, основные потребители кредитов - люди от 30 до 60 лет. Пропуски есть во всех группах,
#в которых люди берут кредиты.
#Посмотрим, какой возраст встречается чаще всего, заодно проверив предыдущий вывод
print('В каком возрасте люди берут кредиты чаще всего?')
print(data.groupby('dob_years')['dob_years'].count().sort_values(ascending=False).head(5))
print()
print()
#посмотрим максимальное и минимальное значение зарплат среди заемщиков
print('Максимум и минимум при заёме:')
print('Максимальная зарплата при заёме:',data['total_income'].max())
print('Минимальная зарплата при заёме:',data['total_income'].min())
#разница между минимальным и максимальными значениями слишком большая, чтобы считать средние или медианы по
#всем возрастам, поэтому воспользуемся уже существующей градацией по возрастам. Посчитаем средние зарплаты и медианы в 
#соответствующих группах и заполним пропуски характерными значениями.
print()

#поскольку значения графы total_income являются количественными, нам нужно заполнить пропуски характерными значениями.
#характерные значения могут принимать вид средних или медиан, в зависимости от того какой параметр будет являться более точным
#Есть гипотеза, что чем старше люди, чем сильнее будет разброс в деньгах в месяц. Поэтому создадим таблицу, из 5 столбцов:
#возратной интервал, максимальная зарплата, минимальная зарплата, средняя, медиана.Чтобы наглядно можно было оценить какой параметр подйдет лучше.

#напишем функцию, которая будет считать все 4 параметра и выводить их в одной строке. Входные переменные - рамки возраста
#выходные: значения максимума, минимума, среднего и медианы
def find_best_salary(age1,age2):
    
    max1 = data['total_income'][(data['dob_years']>=age1) & (data['dob_years']<=age2)].max()
    min1 = data['total_income'][(data['dob_years']>=age1) & (data['dob_years']<=age2)].min()
    mean1 = data['total_income'][(data['dob_years']>=age1) & (data['dob_years']<=age2)].mean()
    median1 = data['total_income'][(data['dob_years']>=age1) & (data['dob_years']<=age2)].sort_values(ascending = False).median()
    column = pd.Series([max1,min1,mean1,median1], index=['Максимум','Минимум','Средняя','Медиана'])
    return column
    
salary1_20 = find_best_salary(1,20)
salary21_30 = find_best_salary(21,30)
salary31_40 = find_best_salary(31,40)
salary41_50 = find_best_salary(41,50)
salary51_60 = find_best_salary(51,60)
salary61_70 = find_best_salary(61,70)
salary71_100 = find_best_salary(71,100)
main_salary_grouped = pd.concat([salary1_20, salary21_30, salary31_40, salary41_50, salary51_60, salary61_70, salary71_100], axis =1)
main_salary_grouped.columns = ['1 до 20 лет', '21 до 30 лет','31 до 40 лет', 'от 41 до 50 лет', 'от 51 до 60 лет', 'от 61 до 70 лет', 'от 71 до 100 лет']
print()
print(main_salary_grouped)
print()
print()

#заменяем пропущенные значения на средние по группам. Напишем функцию, которая будет выполнять это за нас

def change_salary(data,age1,age2):
    data['total_income'][(data['dob_years']>=age1) & (data['dob_years']<=age2)] = data['total_income'][(data['dob_years']>=age1) & (data['dob_years']<=age2)].fillna(data['total_income'][(data['dob_years']>=age1) & (data['dob_years']<=age2)].mean())
                                                                        

change_salary(data,1,20) 
change_salary(data,21,30)
change_salary(data,31,40)
change_salary(data,41,50)
change_salary(data,51,60)
change_salary(data,61,70)
change_salary(data,71,100)

print()
print()
#data[data['days_employed']['days_employed']<0] = abs(data[data['days_employed']['days_employed']<0])
data['days_employed'][data['days_employed']<0] = abs(data['days_employed'][data['days_employed']<0]) #убираем отрицательные
#значения в столбце 'days_employed'.

print()
print()
print('Максимальный стаж при заёме:',data['days_employed'].max())
print('Минимальный стаж при заёме:',data['days_employed'].min())
#избавляемся от аномально высоких стажей
data['days_employed'][data['days_employed']>18250] = data['days_employed']/24
#создаем массив с целыми числами от 0 до 100, которые потом будем подставлять в цикл для заполнения пропущенных значений медианами
years_list = list(range(0, 100))
def days_employed_median(age):
    data['days_employed'][data['dob_years'] == age] = data['days_employed'][data['dob_years'] == age].fillna(data['days_employed'][data['dob_years'] == age].sort_values(ascending=False).median())

for i in years_list:
    days_employed_median(i)

print()
print()
print(data.info()) #проверяем отсутствие пропусков. Видим 10 пропусков. Это пропуски в тех строках, где пропущенно и значение возраста
#и сумма по зарплате. Точные значения по зарплате подставить там не удастся, поэтому подставим в них медианные значения по всей таблице.
data['total_income'][data['dob_years'] == 0] = data['total_income'][data['dob_years'] == 0].fillna(data['total_income'].sort_values(ascending=False).median())







Ищем пропуски в столбцах
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


Убеждаемся в отсутствии минусовых значений зарплат
0


Считаем, сколько заемщиков берут средства в разных интервалах
    Возраст заемщиков  Их количество  Количество пропусков
0        1 до 20 лет:             65                     6
1       21 до 30 лет:           3658                   351
2       31 до 40 лет:           5743                   573
3    от 41 до 50 лет:           5276                   522
4    от 51 до 60 лет:           4542                   489
5    от 61 до 70 лет:           2034                   215
6   от 71 до 100 лет:            106                     8
7  от 100 до 200 лет:              0                     0

В каком воз

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:93: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:107: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:115: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:119: SettingWithCopyWarning: 
A value is trying to b





Максимальный стаж при заёме: 401755.40047533
Минимальный стаж при заёме: 24.14163324048118


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21515 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:128: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


### Вывод

Более тщательная проверка на пропуски с помощью метода isnull() подтвердила гипотезу о пропусках в столбцах days_employed и total_income. Первым делом мы закрыли пропуски в столбце 'total_income'. Поскольку таблица содержит в себе данные людей разного возраста, семейного положения, образования и т.д. некорректно было бы просто считать средние или медианные значения по всей таблице.
Поэтому я решил объединить участников в группы по возрастам, для каждой группы посчитал максимум, минимум, средние, медианы. 
Но сперва я обнаружил, что 101 строка содержит в столбце возраст значение 0, что невозможно. Пропуски можно объяснить несколькими причинами:
1. Возраст людей равен числу, кратному 10 (20,30,40,50 и т.п.) в таком случае сотрудница банка могла просто машинно ошибиться, не пропечатав нужную цифру.
2. Данные были не нужны/важны для цели кредита (кредит выдавался на незначительную сумму), поэтому возраст не являлся обязательным параметром, поле оставили незаполненным)
3. Потеря данных при переводе из формата в формат. Данные могли быть заполнены в виде "24 года" и при изымании данных из базы, автоматически проставились 0 в этих ячейках.

Поскольку неясно к какой категории принадлежат заемщики с цифрой 0 в графе возраст, использовать их данные по зарплатам нельзя.  Cредние и медианы посчитаем по тем данным, где возраст указан.
Поскольку большой разницы между средними и медианами обнаружено не было - я взял значения средних по группам. Пропуски в графе total_income объяснить сложно, поскольку это самы важный столбец в анализе выдачи кредита и для решении по выдаче оного. Могу объясниь только потерей данных при экспорте таблицы из баз данных, в связи с первоначальным некорректным заполнением.
Оставшиеся 10 строк с пропусками и в возрасте и в зарплате заполнил медианными значениями по всей таблице.

Со столбцом days_employed все оказалось сложнее. В нем встречаются, как отрицательные значения, так и неадекватные(Максимальный стаж при заёме: 401755, что равняется 1100 лет, а это начало карьеры при Игоре Рюриковиче). Значения отрицательные еще можно объяснить путаницей со знаком "-" при заполнении. Аномальные значения стажа только тем, что напутаны единицы измерения. Возможно, это подсчитаны часы работы, а не года. В любом случае, нет задачи строить выводы относительно данного столбца, поэтому отрицательные значения исправим с помощью функции модуля abs(). Значения, которые отражают стаж более 50 лет разделим на количество часов в сутках, чтобы получить предположительные года. А пропуски после этого заполним медианными значениями по каждому возрасту. Если бы требовался детальный анализ, связанный с данным столбцом - необходимо получить комментарии по нему у сотрудника, что составлял таблицу.
В конце вызываем метод info для проверки пропусков, коих не обнаруживаем.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Очень основательный подход к заполнению пропусков, как вариант также можно было провести группировку по типу занятости клиента. Такой большой и развернутый вывод, редко такое встретишь!)
    
</div>

<div class="alert alert-block alert-warning">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

1. Нулевой возраст вероятнее всего маркер пропуска, можно как вариант заполнить его путем группировки по образованию

2. Посмотри на уникальные значения пола, там тоже есть интересный пропуск
    
</div>

### Замена типа данных

In [28]:
#займемся детьми. Посмотрим, есть ли отрицательные значения и слишком большие значения количества детей
print('Есть ли отрицательные значения количества детей?')
print(data['children'][data['children']<0].count()) #отрицательные значения есть. Возьмем их модуль и убирем отрицательные.
data['children'][data['children']<0] = abs(data['children'][data['children']<0]) #убираем отрицательные
print('Проверка количества детей')
print(data.groupby('children')['children'].count()) #75 значений с показателем 20, что явно ошибка, и скорее всего кратная 10
# разделим эти значения на 10
print()
print()
print('Повторная проверка количества детей')
data['children'][data['children'] == 20] = 2 
print(data.groupby('children')['children'].count())
# c детьми разобрались

#заменим типы данных float на целочисленные. Для этого сначала снова вызовем info,чтобы посмотреть столбцы, требующие замены
print(data.info()) 
#замены требуют total_income и days_employed
data['total_income'] = data['total_income'].astype('int64') #заменяем на int
data['days_employed'] = data['days_employed'].astype('int64') #заменяем на int
print()
print()




Есть ли отрицательные значения количества детей?
47
Проверка количества детей
children
0     14149
1      4865
2      2055
3       330
4        41
5         9
20       76
Name: children, dtype: int64


Повторная проверка количества детей
children
0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory u

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


### Вывод

Прежде чем менять типы данных, я решил изучить остальные столбцы, в частности детский.Подсчет отрицательных значений дал результат 47. Скорее всего ошибка при заполнении, связанная с "тире", берем в модуль их. 
Еще один занимательный артефакт - 75 заемщиков с 20ю детьми. Семьи с таким количеством детей конечно существуют, но сомнительно, что все они пришли в один и тот же банк. Эти значения поделим на 10, и оставим в таблице их как двудетных заемщиков. Ошибка, возможно, перекочевала от числа с плавающей точкой 2.0. 
Метод info показал нам, что два столбца имеют типы float64. Поскольку все операции с ними мы уже выполнили, убрали пропуска в них, ничто нам не мешает привести их к более человеческому и наглядному типу int64. Для этого я воспользовался функцией astype. Почему? я другой не знаю. И не нашел при первом приближении.


<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Все очень даже хорошо! astype тут собственно как нельзя кстати)
    
</div>

### Дубликаты

In [29]:
#прежде чем искать дубликаты, нам нужно посмотреть, все ли правильно с регистром и ошибками в строковых столбцах, для этого вызовем метод
#unique и посмотрим на результаты.

print(data['education'].unique())
print(data['family_status'].unique())
print(data['gender'].unique())
print(data['income_type'].unique())

data['education'] = data['education'].str.lower()
print(data['education'].unique())

#Посчитаем, есть ли дубликаты в нашей таблице
print('Количество дубликатов в таблице:',data.duplicated().sum()) #в таблице 71 дубликата. Удалим их
data = data.drop_duplicates().reset_index(drop= True)

#в процессе поиска уникальных значений мы обнаружили артефакт в виде пола XNА
print(data['gender'][data['gender'] == 'XNA'].count())
#посчтив количество данных значний, мы выясняем, что всего один такой заемщик. Поскольку в общей задаче не стоит задания выявить 
#взаимосвязи погашения долга и пола - заменим это значение на наиболее популярный пол среди заемщиков.
print()
print()
print('Посчитаем какой пол преобладает при заёмах:')
print(data['gender'].value_counts())
#
data['gender'][data['gender'] == 'XNA'] = 'F'
print()
print()
print(data['gender'].value_counts())




['высшее' 'среднее' 'Среднее' 'СРЕДНЕЕ' 'ВЫСШЕЕ' 'неоконченное высшее'
 'начальное' 'Высшее' 'НЕОКОНЧЕННОЕ ВЫСШЕЕ' 'Неоконченное высшее'
 'НАЧАЛЬНОЕ' 'Начальное' 'Ученая степень' 'УЧЕНАЯ СТЕПЕНЬ'
 'ученая степень']
['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем']
['F' 'M' 'XNA']
['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете']
['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']
Количество дубликатов в таблице: 71
1


Посчитаем какой пол преобладает при заёмах:
F      14174
M       7279
XNA        1
Name: gender, dtype: int64


F    14175
M     7279
Name: gender, dtype: int64


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


### Вывод

Искать дубликаты среди числовых столбцов - нет смысла. Друго дело, посмотреть на значения строковых столбцов и поискать там различные ошибки.
Проверка по уникальным значениям ясно дала понять, что не в порядке только столбец с образованием, из-за проблем с регистром. 
Регист ровняем по меньшем. И дальше таблица готова к глобальному поиску дубликатов с помощью duplicated().sum(). Этим методом мы нашли 71 дубликат, с которыми можно смело попрощаться с помощью метода drop_duplicates(). 
Дубликаты могли возникнуть из-за:
1. Стыковка различных баз друг с другом.
2. Ошибки в CRM системах, повторное создание карточек клиентов.
3. Заемщик переоформлял кредиты с помощью других кредитов и т.п.

При просвечивании на "уникальность" мы обнаружили интересный пол у одного заемщика - XNA. Согласно последним трендам, в банк действительно могло войти существо с неопределнным полом и сотрудник банка просто растерялся и оставим поле пустым. Но скорее всего, это просто ошибка сотрудника банка, который не заполнил поле или заполнил его некорректно. Значение всего одно, поэтому поставим вместо него пустую строку.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Молодец, с категориальными признаками поработал, дубликаты удалил
    
</div>
<div class="alert alert-block alert-warning">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Строку с XNA лучше удалить вовсе, либо заполнить наиболее популярным полом. Признак пола как вариант может пригодиться при дополнительно группировке в блоке вопросов и значение XNA будет только мешать (твое заполнение пустым место по сути картины не меняет)
    
</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 

Молодец, доработал!

</div>

### Лемматизация

In [30]:
from pymystem3 import Mystem
m = Mystem()

def lemmatize(data):
    return m.lemmatize(data)
lemmas = data['purpose'].apply(lemmatize)
data['lemma'] = lemmas
print(data['lemma'].value_counts())









[автомобиль, \n]                                          972
[свадьба, \n]                                             791
[на,  , проведение,  , свадьба, \n]                       768
[сыграть,  , свадьба, \n]                                 765
[операция,  , с,  , недвижимость, \n]                     675
[покупка,  , коммерческий,  , недвижимость, \n]           661
[операция,  , с,  , жилье, \n]                            652
[покупка,  , жилье,  , для,  , сдача, \n]                 651
[операция,  , с,  , коммерческий,  , недвижимость, \n]    650
[покупка,  , жилье, \n]                                   646
[жилье, \n]                                               646
[покупка,  , жилье,  , для,  , семья, \n]                 638
[строительство,  , собственный,  , недвижимость, \n]      635
[недвижимость, \n]                                        633
[операция,  , со,  , свой,  , недвижимость, \n]           627
[строительство,  , жилой,  , недвижимость, \n]            624
[покупка

### Вывод

Проведем лемматизацию столбца purpose и результаты выведем в отдельный столбец lemmа, попутно состчитав, что получилось в итоге. Глянув на все цели кредита, можно сделать вывод, что все несколько целей описаны по-разному и многократно. Выделим из всех целей несколько основных: недвижимость, автомобиль, ремонт, образование, свадьба. "Недвижимость" вполне логично разбить на две части: обычная и коммерческая, поскольку на практике это совсем разные кредиты, как например на покупку квартиры семье или для покупку складских помещений под аренду. 

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Молодец, ты хорошо провел лемматизацию и выделил наиболее популярные цели кредита
    
</div>

### Категоризация данных

In [31]:
def purpose_type(text):
    if 'ремонт' in text:
        return 'ремонт'
    elif 'коммерческий' in text:
        return 'коммерческая недвижимость'
    elif 'автомобиль' in text:
        return 'автомобиль'
    elif 'свадьба' in text:
        return 'свадьба'
    elif 'недвижимость' in text:
        return'недвижимость'
    elif 'жилье' in text:
        return'недвижимость'
    elif 'образование' in text:
        return 'образование' 
    
print()
print()    
data['true_purpose'] = data['lemma'].apply(purpose_type)
print('На что берут кредиты:')
print(data['true_purpose'].value_counts())
print()
print()
print('Посмотрим на распределение зарплат:')
print(data['total_income'].describe())

def total_income_difference(income):
    if income < 107623:
        return 'низкий'
    elif income > 107623 and income < 153861:
        return 'средний'
    elif income > 153861 and income < 195813:
        return 'высокий'
    elif income > 195813:
        return 'мажоры'
data['total_income_grouped'] = data['total_income'].apply(total_income_difference)
print()
print()  
print('Уровень дохода при кредитовании, 4 равных категории:')
print(data['total_income_grouped'].value_counts())
print()
print() 
def children_difference(income):
    if income < 1:
        return 'нет детей'
    elif income > 0 and income < 3:
        return 'малодетная семья'
    elif income > 2:
        return 'многодетная семья'
data['children_grouped'] = data['children'].apply(children_difference)
print('Посмотрим на результаты категоризации по детям')
print(data[['children','children_grouped']].head(10))




На что берут кредиты:
недвижимость                 8893
автомобиль                   4306
образование                  4013
свадьба                      2324
коммерческая недвижимость    1311
ремонт                        607
Name: true_purpose, dtype: int64


Посмотрим на распределение зарплат:
count    2.145400e+04
mean     1.674143e+05
std      9.785933e+04
min      2.066700e+04
25%      1.076230e+05
50%      1.538615e+05
75%      1.958132e+05
max      2.265604e+06
Name: total_income, dtype: float64


Уровень дохода при кредитовании, 4 равных категории:
низкий     5364
мажоры     5364
высокий    5363
средний    5363
Name: total_income_grouped, dtype: int64


Посмотрим на результаты категоризации по детям
   children   children_grouped
0         1   малодетная семья
1         1   малодетная семья
2         0          нет детей
3         3  многодетная семья
4         0          нет детей
5         0          нет детей
6         0          нет детей
7         0          нет детей
8 

<div class="alert alert-block alert-warning">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

В коде дублирование блока для среднего дохода
    
</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 
    
Good, подправил

</div>

### Вывод

Возьмем глобальные цели кредитов, которые мы определили при лемматизации и создадим новый столбец true_purpose, который будет отражать их.  Сделав подсчет, мы выявили, что чаще всего кредиты берут для покупки или операциями с жилой недвижимостью, затем следуют автокредиты.

Поскольку одной из наших задач является выявление зависимости возврата кредитов от уровня дохода, эти самые уровни необходимо определить. С помощью метода describe() я равноценно разбил все категории дохода заемщиков.

Тем же методом, что и со столбцом true_purpose, создаем новый столбец, куда записываем соответствующие уровни дохода.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Ахахахах))) 

1. `Доход`

Интересный и хороший способ категоризации по доходу, как вариант можно было воспользоваться и таким подходом:
    
*  Вывести статистики с помощью describe, так бы сразу все было видно

* Либо построить ящик с усами https://plotly.com/python/box-plots/
    
*  У нас квартилями распределение очень хорошо делится на 25%. Иными словами, все что до 1 квартиля (25%) — низкий доход, между 1 и мединой(2 квартиль) тоже будет 25% — средний доход, от медианы до 3 квартиля (25%) — выше среднего, а все что дальше 3 квартиля — высокий доход.
    
2. `Дети`
    
Я бы еще категоризировал по кол-ву детей, это не критично, но было бы удобно:
    
* Традиционные группы «без детей», «малодетная» (1-2 ребенка), «многодетная» (3-...)
 
* И «есть дети \ нет детей»
</div>

### Шаг 3. Ответьте на вопросы

<div class="alert alert-block alert-danger">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Думаю гораздо оптимальнее и нагляднее считать % невозвратов делив должников на общее число клиентов по группе, так различия будут более явными, можешь добавить этот признак в свои таблички
    
</div>

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [32]:
children_depending = data.pivot_table(index = ['children'], columns = 'debt', values= 'total_income', aggfunc='count')
children_depending = children_depending.fillna(0)

children_depending['percent'] = children_depending[1] / (children_depending[0] + children_depending[1])
print(data.info())
print(children_depending.sort_values('percent',ascending=False))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 16 columns):
children                21454 non-null int64
days_employed           21454 non-null int64
dob_years               21454 non-null int64
education               21454 non-null object
education_id            21454 non-null int64
family_status           21454 non-null object
family_status_id        21454 non-null int64
gender                  21454 non-null object
income_type             21454 non-null object
debt                    21454 non-null int64
total_income            21454 non-null int64
purpose                 21454 non-null object
lemma                   21454 non-null object
true_purpose            21454 non-null object
total_income_grouped    21454 non-null object
children_grouped        21454 non-null object
dtypes: int64(7), object(9)
memory usage: 2.6+ MB
None
debt            0       1   percent
children                           
4            37.0     4.0  0.097561

### Вывод

Как показывает сводная таблица, лучше всех выплачивают семьи с 5ю детьми. Поскольку выборка слишком мала, то сделаем вывод о том, что лучше всех справляются с кредитами заемщики без детей, оно и понятно, меньше трат, иногда непредвиденных. 
Немного выбивается из общей картины семьи с 3 детьми

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 

Молодец, вывод хороший. Можно было еще домножить на 100 признак percent, чтобы воспринималось еще лучше)

</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Подсчитай по % невозврата, в целом наблюдаем тенденцию, что дети негативно влияют на возврат кредита
    
</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [33]:
family_status_depending = data.pivot_table(index = ['family_status'], columns = 'debt', values= 'total_income', aggfunc='count')
family_status_depending = family_status_depending.fillna(0)

family_status_depending['percent'] = family_status_depending[1] / (family_status_depending[0] + family_status_depending[1])
print(family_status_depending.sort_values('percent',ascending=False))
print()
print()
family_status_depending1 = data.pivot_table(index = 'true_purpose', columns = 'family_status', values= 'total_income', aggfunc='count')
print('На что берут кредиты вдовцы и вдовы:')
print(family_status_depending1['вдовец / вдова'])

debt                       0    1   percent
family_status                              
Не женат / не замужем   2536  274  0.097509
гражданский брак        3763  388  0.093471
женат / замужем        11408  931  0.075452
в разводе               1110   85  0.071130
вдовец / вдова           896   63  0.065693


На что берут кредиты вдовцы и вдовы:
true_purpose
автомобиль                   218.0
коммерческая недвижимость     77.0
недвижимость                 433.0
образование                  199.0
ремонт                        32.0
свадьба                        NaN
Name: вдовец / вдова, dtype: float64


### Вывод

Сводная таблица показывает, что лучше всего возвращают кредиты вдовцы и вдовы, хуже всего не обремененные ответственностью холостые люди. Возможно это можно объяснить, что вдовцы и вдовы имеют часть имущества от умершего супруга, поэтому изначально входят в кредит не с пустыми руками. Ожидаемо на свадьбы они не тратятся.
Чаще услугами кредитов пользуются женатые или замужние заемщики. Тоже логично, им покупать квартиры, машины для семьи.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 

Молодец, вывод хороший. Видим явное влияние статуса официальности отношений на возврат кредита (как вариант можно было провести дополнительную группировку по полу)

</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Да, действительно люди, которые грубо говоря не состоят \ не состояли в официальных отношениях хуже возвращают кредиты. Вот тут хорошо было бы проводить дополнительную группировку по полу.
    
</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
total_income_grouped_depending = data.pivot_table(index = ['total_income_grouped'], columns = 'debt', values= 'total_income', aggfunc='count')
total_income_grouped_depending = total_income_grouped_depending.fillna(0)

total_income_grouped_depending['percent'] = total_income_grouped_depending[1] / (total_income_grouped_depending[0] + total_income_grouped_depending[1])
print(total_income_grouped_depending.sort_values('percent',ascending=False))

debt                     0    1   percent
total_income_grouped                     
высокий               4895  468  0.087265
средний               4900  463  0.086332
низкий                4937  427  0.079605
мажоры                4981  383  0.071402


### Вывод

Среди заемщиков наблюдается тенденция о том, что более надежными являются клиенты с доходами низкого уровня и мажоры. Средний и высокий уровень оказались менее надежными для банков заемщиками.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 

Я оценил твое чувство юмора, но все же не стоит так называть категории людей, как «мажоры» (вдруг там селфмейды, обидятся хаха)

По поводу выводов, да, удивительно, но люди среднего дохода менее надежны

</div>

<div class="alert alert-block alert-warning">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Среди?)) Так так и что там у нас?) Мы видим, что наиболее рисковая группа это люди среднего дохода, тогда как клиенты с низким и высоким доходом надежнее
    
</div>

- Как разные цели кредита влияют на его возврат в срок?

In [19]:
purpose_depending = data.pivot_table(index = ['true_purpose'], columns = 'debt', values= 'total_income', aggfunc='count')
purpose_depending = purpose_depending.fillna(0)

purpose_depending['percent'] = purpose_depending[1] / (purpose_depending[0] + purpose_depending[1])
print(purpose_depending.sort_values('percent',ascending=False))

debt                          0    1   percent
true_purpose                                  
автомобиль                 3903  403  0.093590
образование                3643  370  0.092200
свадьба                    2138  186  0.080034
коммерческая недвижимость  1212   99  0.075515
недвижимость               8245  648  0.072866
ремонт                      572   35  0.057661


### Вывод

Ожидаемо целью - лидером по кредитованию стала недвижимость. Так как это самая трудная цель для накопления средств, а жить где-то нужно. Приятно, что образование совсем немного отстало от автомобилей и обыватели понимаю пользу от образования и даже берут кредиты для его получения. 
При этом если смотреть на наличие задолженностей, то мы видим, что автокредиты имеют наименьшую возвратность, большую недвижимость и ремонт. Связано это скорее всего с серьезностью данных кредитов. Пользователи автокредитов не всегда понимают дополнительные расходы, которые могут быть связаны с обслуживанием автомобиля,особенно если берут подежаные версии авто. 

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА 2:</b> 

Хороший вывод!

</div>

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Да, действительно автомобили и образование наиболее рисковые цели, тогда как недвижимость и свадьба более надежны
    
</div>

<div class="alert alert-block alert-danger">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Пожалуйста подсчитай % невозврата, так будут лучше видны различия
    
</div>

### Шаг 4. Общий вывод

Исследование показало нам следующие основные критерии, связанные со статистикой по кредитам. Ему были подвержены основные критерии, по которым банк оценивает заемщиков (возраст, наличие детей, доход, цели кредита и т.д.)
На этапе предобработки данных мы столкнулись с такими проблемами как:
- отсутствие данных в графе о доходах
- отсутствие и неадекватные данные в графе об общем трудовом стаже
- минусовые значения в грарфе количества детей
- нулевые значения возраста
- наличие дубликатов в графе образование и в целом по таблице
- задублированные цели кредита

Решив все задачи предобрабоки данных можно сделать слудющие глобальные выводы:
- несмотря на разницу в зависимости наличия задолженностей, кредиты возвращаяются в 90% случаем минимум. т.е. банк может рассчитывать на то, 9 из 10 заемщиков вернут средства вовремя.
- меньше всего задолженностей связаны с целями на недвижимость (покупка, операции, ремонт). Заемщики данныхк кредитов чаще возвращают деньги вовремя. Хуже всего дела обстоят с автокредитами. Разница составляет 2%, поэтому для продолжения исследдования следует посмотреть на суммы задолженностей. Поскольку один невозврат по недвижимости может быть выше, чем несколько невозвратов по автокредитованию.
- заемщики без детей возвращают кредиты лучше. Чем больше детей, тем выше вероятность непредвиденных расходов на одежду, медицину и т.п., что ведет к появлению задолженностей. Удевительным выглядит наличие заемщиков с 3 детьми вверху списка. Объяснить это логически не получается.
- чем выше достаток людей, тем лучше они возвращают деньги. Исключением стали лишь заемщики с низким уровнем дохода, данную аномалию можно объяснить искусственным заполненением 2174 поля в столбце 'total_income'. Пропуски в полях данной информации существенно влияют на статистику, особенно если учитывать количество данных пропусков.
- лучше всего возвращают кредиты вдовцы и разведенные заемщики. Скорее всего потому, что это уже люди в возрасте, состоявшиеся, берут кредит спланированно, обдуманно. Хуже всего дела у неженатых и незамужних, заемщики данной категории явно моложе, ответственности у них меньше.
 

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Хорошие выводы, ты большой молодец!
    
</div>

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.

<div class="alert alert-block alert-success">
<b>КОММЕНТАРИЙ РЕВЬЮВЕРА:</b> 

Сергей, мне очень даже понравился твой проект, ты большой молодец, есть буквально несколько моментов на правку!)
    
</div>